Classes/SegmentService.swift

```diff
@@ -49,25 +52,25 @@ public class SegmentService {
         completion(log)
     }
     
-    public func segmentCallsIn(from log: [ProxyLogElement], completion: @escaping ([Segment]) -> Void) {
-        var segmentList: [Segment] = []
+    public func segmentCallsIn<T: SegmentBatchCodable>(from log: [ProxyLogElement], completion: @escaping ([Segment<T>]) -> Void) {
+        var segmentList: [Segment<T>] = []
         for element in log {
             if element.host == "api.segment.io" {
                 let segmentString = element.request?.body?.text
                 let segmentData = segmentString!.data(using: .utf8)!
-                let segment = try! self.decoder.decode(Segment.self, from: segmentData)
+                let segment = try! self.decoder.decode(Segment<T>.self, from: segmentData)
                 segmentList.append(segment)
             }
         }
         completion(segmentList)
     }
     
-    public func matchingSegmentBatchesIn(completion: @escaping ([BatchElement]) -> Void, from segmentList: [Segment], expectedCallType: String, expectedPageType: String) {
-        var expectedBatchElements: [BatchElement] = []
+    public func matchingSegmentBatchesIn<T: SegmentBatchCodable>(completion: @escaping ([T]) -> Void, from segmentList: [Segment<T>], expectedCallType: String, expectedPageName: String) {
+        var expectedBatchElements: [T] = []
         for segmentElement in segmentList {
             let batch = segmentElement.batch
             for batchElement in batch ?? [] {
-                if batchElement.type == expectedCallType && batchElement.properties?.pageType == expectedPageType {
+                if batchElement.type == expectedCallType && batchElement.name == expectedPageName {
```

Comment:
this change from using page type to name is optional but I recommend it unless you prefer the page type for a particular reason.
1. page type can be the same on multiple pages. example in fsapp is all the settings pages are page type "settings-drawer" so this is just a little more specific
2. most importantly here, it relieves some complexity with the generics. I could still use page type, but I'd have to go down a level into the properties. not unreasonable i think, just a little messier in my opinion